compiler/passes/checkResolved.cpp

```diff
@@ -183,14 +187,62 @@ isDefinedAllPaths(Expr* expr, Symbol* ret, RefSet& refs)
   if (TryStmt* tryStmt = toTryStmt(expr))
   {
     int result = INT_MAX;
-    for_alist(c, tryStmt->_catches)
+
+    // This indicates whether or not we find a catch-all case, in
+    // which case nothing can escape us unless the individual clauses
+    // let it; if this is a try! statement, it doesn't need a
+    // catch-all case, so set it to true
+    //
+    bool foundCatchall = tryStmt->tryBang();
+    for_alist(c, tryStmt->_catches) {
       result = std::min(result, isDefinedAllPaths(c, ret, refs));
+      if (toCatchStmt(c)->isCatchall()) {
+        foundCatchall = true;
+      }
+    }
 
-    return std::min(result, isDefinedAllPaths(tryStmt->body(), ret, refs));
+    result = std::min(result, isDefinedAllPaths(tryStmt->body(), ret, refs));
+
+    // even if the try and all catches are air-tight, if there's no
+    // catch-all, we can escape via an uncaught error, and will need
+    // for our parent statement to contain returns as well...
+    if (result == 1 && !foundCatchall) {
+      result = 0;
+    }
+    return result;
   }
 
-  if (CatchStmt* catchStmt = toCatchStmt(expr))
-    return isDefinedAllPaths(catchStmt->body(), ret, refs);
+  if (CatchStmt* catchStmt = toCatchStmt(expr)) {
+    if (catchStmt->isCatchall()) {
+      // if this is a catch-all, the answer is whatever its body says
+      return isDefinedAllPaths(catchStmt->body(), ret, refs);
+    } else {
+      // otherwise, this is a case that (unfortunately) looks like this:
+      //
+      // 'if (errorMatches) { user code } else { /* nothing */ }
+      //
+      // where that else clause doesn't reflect the true control flow,
+      // so we need to special-case it (later the next catch block
+      // will be inlined there).  Open Q: Why isn't this done
+      // outright?
+
+      // Find the else body in the last CondStmt in the catch block body.
+      // This should always exist after CatchStmt::cleanup
+      // for non-catchall errors.
+
+      CondStmt* finalCond = NULL;
+      for_alist_backward(node, catchStmt->body()->body) {
```

Comment:
I'd also / alternativey consider adding a flag to mark the errorMatches VarSymbol .

Reply:
Are you saying you'd suggest doing this when we build up the conditional in the first place to avoid having to search for / assume a particular control flow structure in the catch blocks?

Reply:
yes, exactly. you'd search for the `if ( var with special flag )` instead. But I think this is OK for the moment.